staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,617 @@
+		})
+		return err
+	}
+
+	// Check cache entry for this groupversion and insert it into the
+	// document if present
+	if entry, exists := cached.discovery[mgv]; exists {
+		//!TODO: mark with staleness of cache
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, entry)
+	} else {
+		// Use empty GV, since there is an APIService for it
+		//!TODO: mark with staleness of cache
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, apidiscoveryv2beta1.APIVersionDiscovery{
+			Version: gv.Version,
+		})
+	}
+
+	return nil
+}
+
+// Spwans a goroutune which waits for added/updated apiservices and updates
+// the discovery document accordingly
+func (dm *discoveryManager) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting ResourceDiscoveryManager")
+
+	// Spawn workers
+	// These workers wait for APIServices to be marked dirty.
+	// Worker ensures the cached discovery document hosted by the ServiceReference of
+	// the APIService is at least as fresh as the APIService, then includes the
+	// APIService's groupversion into the merged document
+	for i := 0; i < 2; i++ {
+		go func() {
+			for {
+				next, shutdown := dm.dirtyAPIServiceQueue.Get()
+				if shutdown {
+					return
+				}
+
+				if err := dm.syncAPIService(next.(string)); err != nil {
+					dm.dirtyAPIServiceQueue.AddRateLimited(next)
+				} else {
+					dm.dirtyAPIServiceQueue.AddAfter(next, 5*time.Second)
+				}
+				dm.dirtyAPIServiceQueue.Done(next)
+			}
+		}()
+	}
+	<-stopCh
+
+	// Shutdown the queue since stopCh was signalled
+	dm.dirtyAPIServiceQueue.ShutDown()
+}
+
+// Adds an APIService to be tracked by the discovery manager. If the APIService
+// is already known
+func (dm *discoveryManager) AddAPIService(apiService *apiregistrationv1.APIService, handler http.Handler) {
+	dm.servicesLock.Lock()
+	defer dm.servicesLock.Unlock()
+
+	// If service is nil then its information is contained by a local APIService
+	// but we can not know which.
+	if apiService.Spec.Service == nil {
+		gv := helper.APIServiceNameToGroupVersion(apiService.Name)
+
+		// If the local delegate responds to /apis/<group>/version
+		//	or /api/<version> (for empty group)
+		// then we will use that as a service
+		searchPath := "/apis/" + gv.Group + "/" + gv.Version
+		if len(gv.Group) == 0 {
+			searchPath = "/api/" + gv.Version
+		}
+
+		i := 0
+		for name, delegate := range dm.localDelegates {
+			i += 1
+			listedPaths := delegate.ListedPaths()
+			for _, path := range listedPaths {
+				// TODO(jefftree): Magic to obtain the CRD handler, this needs to be refactored.
+				if path == searchPath || i == 2 {
+					dm.apiServices[apiService.Name] = groupVersionInfo{
+						handler:         delegate.UnprotectedHandler(),
+						lastMarkedDirty: time.Now(),
+						service: serviceKey{
+							LocalName: name,
+						},
+					}
+					dm.dirtyAPIServiceQueue.Add(apiService.Name)
+					return
+				}
+			}
+		}
+
+		klog.Infof("Failed to find local service for apiservice: %s", apiService.Name)
```

Comment:
if this code fails, then the discovery for the cluster is wrong and it will never recover, right?  If a resource doesn't exist, namespace lifecycle cleanup would start missing cleanups.  This seems a lot more severe than an info.  Fatal perhaps.

Given how I would expect the resourcemanager to work, why do we even have to look things other than remote apiservces ? The CRDs are directly wired and the built-in types can be loaded immediately after installing the apigroup.